File: js/src/builder.ts
/* Builder — a streaming encoder that produces a Blob. */

import { Buf } from "./format.ts";
import { utf8Encode } from "./format.ts";
import * as E from "./encode.ts";
import { Blob } from "./blob.ts";
import { Value } from "./value.ts";

export class Builder {
  _buf: Buf;

  constructor() {
    this._buf = new Buf();
  }

  // ── scalars ───────────────────────────────────────────────────────
  nil(): Builder {
    E.encNil(this._buf);
    return this;
  }
  boolean(v: boolean): Builder {
    E.encBool(this._buf, v);
    return this;
  }
  integer(x: number | bigint): Builder {
    E.encInteger(this._buf, BigInt(x));
    return this;
  }
  unsignedInteger(x: number | bigint): Builder {
    E.encUnsigned(this._buf, BigInt(x));
    return this;
  }
  real(d: number): Builder {
    E.encReal(this._buf, d);
    return this;
  }
  real32(f: number): Builder {
    E.encReal32(this._buf, f);
    return this;
  }
  string(s: string | Uint8Array): Builder {
    E.encString(this._buf, typeof s === "string" ? utf8Encode(s) : s);
    return this;
  }
  binary(data: Uint8Array): Builder {
    E.encBinary(this._buf, data);
    return this;
  }
  ext(typeCode: number, data: Uint8Array): Builder {
    E.encExt(this._buf, typeCode, data);
    return this;
  }

  // ── fixed-width integers ──────────────────────────────────────────
  int8(x: number | bigint): Builder {
    E.encInt8(this._buf, BigInt(x));
    return this;
  }
  int16(x: number | bigint): Builder {
    E.encInt16(this._buf, BigInt(x));
    return this;
  }
  int32(x: number | bigint): Builder {
    E.encInt32(this._buf, BigInt(x));
    return this;
  }
  int64(x: number | bigint): Builder {
    E.encInt64(this._buf, BigInt(x));
    return this;
  }
  uint8(x: number | bigint): Builder {
    E.encUint8(this._buf, BigInt(x));
    return this;
  }
  uint16(x: number | bigint): Builder {
    E.encUint16(this._buf, BigInt(x));
    return this;
  }
  uint32(x: number | bigint): Builder {
    E.encUint32(this._buf, BigInt(x));
    return this;
  }
  uint64(x: number | bigint): Builder {
    E.encUint64(this._buf, BigInt(x));
    return this;
  }

  // ── containers ────────────────────────────────────────────────────
  arrayHeader(count: number): Builder {
    E.encArrayHeader(this._buf, count);
    return this;
  }
  mapHeader(count: number): Builder {
    E.encMapHeader(this._buf, count);
    return this;
  }

  // ── embedding & timestamp ─────────────────────────────────────────
  raw(data: Uint8Array | Blob): Builder {
    this._buf.pushBytes(data instanceof Blob ? data.data() : data);
    return this;
  }
  value(v: Value): Builder {
    E.encodeValue(this._buf, v);
    return this;
  }
  timestamp(sec: number | bigint, nsec: number = 0): Builder {
    E.encTimestamp(this._buf, BigInt(sec), nsec);
    return this;
  }

  // ── finalize ──────────────────────────────────────────────────────
  build(): Blob {
    return new Blob(this._buf.toBytes());
  }

  get length(): number {
    return this._buf.length;
  }

  static quote(v: Value): Blob {
    return new Builder().value(v).build();
  }
}

File: rust/src/builder.rs
//! `Builder` — a streaming encoder that produces a [`Blob`].

use crate::blob::Blob;
use crate::encode as e;
use crate::value::Value;

/// Append msgpack elements in order, then finalise with [`Builder::build`].
#[derive(Default)]
pub struct Builder {
    buf: Vec<u8>,
}

impl Builder {
    pub fn new() -> Builder {
        Builder { buf: Vec::new() }
    }

    // ── scalars ───────────────────────────────────────────────────────
    pub fn nil(&mut self) -> &mut Self {
        e::enc_nil(&mut self.buf);
        self
    }
    pub fn boolean(&mut self, v: bool) -> &mut Self {
        e::enc_bool(&mut self.buf, v);
        self
    }
    pub fn integer(&mut self, x: i64) -> &mut Self {
        e::enc_integer(&mut self.buf, x);
        self
    }
    pub fn unsigned_integer(&mut self, x: u64) -> &mut Self {
        e::enc_unsigned(&mut self.buf, x);
        self
    }
    pub fn real(&mut self, d: f64) -> &mut Self {
        e::enc_real(&mut self.buf, d);
        self
    }
    pub fn real32(&mut self, val: f32) -> &mut Self {
        e::enc_real32(&mut self.buf, val);
        self
    }
    pub fn string(&mut self, s: &str) -> &mut Self {
        e::enc_string(&mut self.buf, s.as_bytes());
        self
    }
    pub fn string_bytes(&mut self, s: &[u8]) -> &mut Self {
        e::enc_string(&mut self.buf, s);
        self
    }
    pub fn binary(&mut self, data: &[u8]) -> &mut Self {
        e::enc_binary(&mut self.buf, data);
        self
    }
    pub fn ext(&mut self, type_code: i8, data: &[u8]) -> &mut Self {
        e::enc_ext(&mut self.buf, type_code, data);
        self
    }

    // ── fixed-width integers ──────────────────────────────────────────
    pub fn int8(&mut self, x: i8) -> &mut Self {
        e::enc_int8(&mut self.buf, x as i64);
        self
    }
    pub fn int16(&mut self, x: i16) -> &mut Self {
        e::enc_int16(&mut self.buf, x as i64);
        self
    }
    pub fn int32(&mut self, x: i32) -> &mut Self {
        e::enc_int32(&mut self.buf, x as i64);
        self
    }
    pub fn int64(&mut self, x: i64) -> &mut Self {
        e::enc_int64(&mut self.buf, x);
        self
    }
    pub fn uint8(&mut self, x: u8) -> &mut Self {
        e::enc_uint8(&mut self.buf, x as u64);
        self
    }
    pub fn uint16(&mut self, x: u16) -> &mut Self {
        e::enc_uint16(&mut self.buf, x as u64);
        self
    }
    pub fn uint32(&mut self, x: u32) -> &mut Self {
        e::enc_uint32(&mut self.buf, x as u64);
        self
    }
    pub fn uint64(&mut self, x: u64) -> &mut Self {
        e::enc_uint64(&mut self.buf, x);
        self
    }

    // ── containers ────────────────────────────────────────────────────
    pub fn array_header(&mut self, count: u32) -> &mut Self {
        e::enc_array_header(&mut self.buf, count);
        self
    }
    pub fn map_header(&mut self, count: u32) -> &mut Self {
        e::enc_map_header(&mut self.buf, count);
        self
    }

    // ── embedding & timestamp ─────────────────────────────────────────
    pub fn raw(&mut self, data: &[u8]) -> &mut Self {
        self.buf.extend_from_slice(data);
        self
    }
    pub fn raw_blob(&mut self, blob: &Blob) -> &mut Self {
        self.buf.extend_from_slice(blob.data());
        self
    }
    pub fn value(&mut self, v: &Value) -> &mut Self {
        e::encode_value(&mut self.buf, v);
        self
    }
    pub fn timestamp(&mut self, sec: i64) -> &mut Self {
        e::enc_timestamp(&mut self.buf, sec, 0);
        self
    }
    pub fn timestamp_ns(&mut self, sec: i64, nsec: u32) -> &mut Self {
        e::enc_timestamp(&mut self.buf, sec, nsec);
        self
    }

    // ── finalize ──────────────────────────────────────────────────────
    /// Finalise into a [`Blob`]. Takes `&self` so it can terminate a method
    /// chain (`Builder::new().integer(1).build()`).
    pub fn build(&self) -> Blob {
        Blob::from_vec(self.buf.clone())
    }
    pub fn len(&self) -> usize {
        self.buf.len()
    }
    pub fn is_empty(&self) -> bool {
        self.buf.is_empty()
    }

    /// One-shot: encode a single [`Value`] into a [`Blob`].
    pub fn quote(v: &Value) -> Blob {
        let mut b = Builder::new();
        b.value(v);
        b.build()
    }
}
